dash/include/dash/algorithm/Copy.h

```diff
@@ -3,17 +3,26 @@
 
 #include <dash/Future.h>
 #include <dash/Iterator.h>
+#include <dash/internal/Config.h>
 #include <dash/iterator/internal/ContiguousRange.h>
 
-#include <dash/algorithm/LocalRange.h>
-
 #include <dash/dart/if/dart_communication.h>
 
 #include <algorithm>
 #include <future>
 #include <memory>
 #include <vector>
 
+
+/*
+ * Control whether dash::copy uses handles to wait for outstanding transfers
+ * to complete. If set to 0, dash::copy will use flush instead to ensure
+ * completion.
+ */
+#if !defined(DASH_COPY_USE_HANDLES)
```

Comment:
Using handles is a trade-off: it allows you to wait for specific transfers instead of all transfers (which may be handy if multiple threads call `dash::copy` or if other transfers are already in flight). As a draw-back, they require memory allocation and might cause noticeable overhead if there many transfers due to short contiguous ranges. I have not done any comparisons, just leaving the option in here to disable the use of handles. You could compile your application using `-DDASH_COPY_USE_HANDLES=0` to disable them.

Reply:
But than, one should be able to make this decision on a call-by-call basis, not all-or-nothing. I'm not convinced, that this is needed right now, so why introduce it with this PR?

Reply:
Why not? Yes, in general controlling that on a call-by-call basis is preferable. Executing policies have been in the ether for so long, I can only guess when they will [vapor|material]ize. Until then, I don't see any harm in having at least a chance to control it globally. If we find that handle-less `std::copy` is more efficient (benchmarks welcome), we can easily change the default. The infrastructure is in place.

Reply:
> benchmarks welcome

Why should I argue with benchmarks, for something which you want to introduce without showing benchmarks in the first place?

Reply:
That was mainly meant as a hint that I only have infinite resources at hand to work on this issue (it's gotten out of control already, really there is no such thing as a "quick fix"). I will make it a template parameter and add a benchmark example though

Reply:
```
----------------------------------------------------------------------------------
-- Runtime arguments
--   -r              20                                      repetitions per round
--   -n              10                                                     rounds
--   -s            4096      matrix size (number of double elements per dimension)
--   -t             512        tile size (number of double elements per dimension)
----------------------------------------------------------------------------------
units, mpi.impl,                          impl,   total [s],    bandwidth [MB/s]
   16,  craympi,              copy_with_handle,  0.05789930,       2318.12350063
   16,  craympi,           copy_without_handle,  0.03827390,       3506.76905149
   16,  craympi,                    copy_async,  0.05626080,       2385.63490032
   16,  craympi,              copy_with_handle,  0.05594790,       2398.97704829
   16,  craympi,           copy_without_handle,  0.03835725,       3499.14887016
   16,  craympi,                    copy_async,  0.05629005,       2384.39525280
   16,  craympi,              copy_with_handle,  0.05596170,       2398.38546720
   16,  craympi,           copy_without_handle,  0.03835130,       3499.69174448
   16,  craympi,                    copy_async,  0.05625060,       2386.06749084
   16,  craympi,              copy_with_handle,  0.05594790,       2398.97704829
   16,  craympi,           copy_without_handle,  0.03837160,       3497.84027771
   16,  craympi,                    copy_async,  0.05622025,       2387.35558807
   16,  craympi,              copy_with_handle,  0.05598375,       2397.44082881
   16,  craympi,           copy_without_handle,  0.03835040,       3499.77387459
   16,  craympi,                    copy_async,  0.05624890,       2386.13960451
   16,  craympi,              copy_with_handle,  0.05601960,       2395.90657556
   16,  craympi,           copy_without_handle,  0.03834610,       3500.16632721
   16,  craympi,                    copy_async,  0.05623290,       2386.81853506
   16,  craympi,              copy_with_handle,  0.05596895,       2398.07478968
   16,  craympi,           copy_without_handle,  0.03835305,       3499.53205808
   16,  craympi,                    copy_async,  0.05626840,       2385.31267994
   16,  craympi,              copy_with_handle,  0.05596455,       2398.26332920
   16,  craympi,           copy_without_handle,  0.03836975,       3498.00892630
   16,  craympi,                    copy_async,  0.05626480,       2385.46529980
   16,  craympi,              copy_with_handle,  0.05601120,       2396.26588968
   16,  craympi,           copy_without_handle,  0.03837580,       3497.45746017
   16,  craympi,                    copy_async,  0.05623485,       2386.73576972
   16,  craympi,              copy_with_handle,  0.05579010,       2405.76245606
   16,  craympi,           copy_without_handle,  0.03836095,       3498.81136937
   16,  craympi,                    copy_async,  0.05623960,       2386.53418588
```

```
----------------------------------------------------------------------------------
-- Runtime arguments
--   -r              20                                      repetitions per round
--   -n              10                                                     rounds
--   -s            4096      matrix size (number of double elements per dimension)
--   -t              64        tile size (number of double elements per dimension)
----------------------------------------------------------------------------------
units, mpi.impl,                          impl,   total [s],    bandwidth [MB/s]
   16,  craympi,              copy_with_handle,  0.46144025,        290.86697140
   16,  craympi,           copy_without_handle,  0.28127320,        477.17922646
   16,  craympi,                    copy_async,  0.45664310,        293.92260170
   16,  craympi,              copy_with_handle,  0.45475685,        295.14174003
   16,  craympi,           copy_without_handle,  0.28064115,        478.25391251
   16,  craympi,                    copy_async,  0.45654145,        293.98804424
   16,  craympi,              copy_with_handle,  0.45498160,        294.99594709
   16,  craympi,           copy_without_handle,  0.28066755,        478.20892725
   16,  craympi,                    copy_async,  0.45660895,        293.94458431
```

Not using handles clearly has an advantage so I made not using them the default.